backend/src/apiserver/resource/resource_manager.go

```diff
@@ -195,616 +510,1147 @@ func (r *ResourceManager) ArchiveExperiment(ctx context.Context, experimentId st
-	var job *model.Job
-	var err error
-	if enabled {
-		job, err = r.checkJobExist(ctx, jobID)
-	} else {
-		// We can skip custom resource existence verification, because disabling
-		// the job do not need to care about it.
-		job, err = r.jobStore.GetJob(jobID)
-	}
+// Fetches execution logs from a pod.
+func (r *ResourceManager) readRunLogFromPod(ctx context.Context, namespace string, nodeId string, follow bool, dst io.Writer) error {
+	logOptions := corev1.PodLogOptions{
+		Container:  "main",
+		Timestamps: false,
+		Follow:     follow,
+	}
+
+	req := r.k8sCoreClient.PodClient(namespace).GetLogs(nodeId, &logOptions)
+	podLogs, err := req.Stream(ctx)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			glog.Errorf("Failed to read logs from pod %v: %v", nodeId, err)
+		}
+		return util.NewInternalServerError(err, "Failed to read logs from pod %v due to error opening log stream", nodeId)
+	}
+	defer podLogs.Close()
+
+	_, err = io.Copy(dst, podLogs)
+	if err != nil && !errors.Is(err, io.EOF) {
+		return util.NewInternalServerError(err, "Failed to read logs from pod %v due to error in streaming the log", nodeId)
+	}
+	return nil
+}
+
+// Fetches execution logs from a archived pod logs.
+func (r *ResourceManager) readRunLogFromArchive(workflowManifest string, nodeId string, dst io.Writer) error {
+	if workflowManifest == "" {
+		return util.NewInternalServerError(util.NewInvalidInputError("Runtime workflow manifest cannot empty"), "Failed to read logs from archive %v due to empty runtime workflow manifest", nodeId)
+	}
+
+	execSpec, err := util.NewExecutionSpecJSON(util.ArgoWorkflow, []byte(workflowManifest))
+	if err != nil {
+		return util.NewInternalServerError(err, "Failed to read logs from archive %v due error reading execution spec", nodeId)
+	}
+
+	logPath, err := r.logArchive.GetLogObjectKey(execSpec, nodeId)
+	if err != nil {
+		return util.NewInternalServerError(err, "Failed to read logs from archive %v", nodeId)
+	}
+
+	logContent, err := r.objectStore.GetFile(logPath)
 	if err != nil {
-		return util.Wrap(err, "Enable/Disable job failed")
+		return util.NewInternalServerError(err, "Failed to read logs from archive %v due to error fetching the log file", nodeId)
+	}
+
+	err = r.logArchive.CopyLogFromArchive(logContent, dst, archive.ExtractLogOptions{LogFormat: archive.LogFormatText, Timestamps: false})
+	if err != nil {
+		return util.NewInternalServerError(err, "Failed to read logs from archive %v due to error copying the log file", nodeId)
+	}
+	return nil
+}
+
+// Creates a recurring run.
+// Note: when creating a recurring run from a manifest, this triggers creation of
+// a new pipeline and pipeline version that share the name, description, and namespace.
+// Manifest's namespace gets overwritten with the job.Namespace if the later is non-empty.
+// Otherwise, job.Namespace gets overwritten by the manifest.
+func (r *ResourceManager) CreateJob(ctx context.Context, job *model.Job) (*model.Job, error) {
```

Comment:
This function also became really long (just like `CreateRun()`) and convoluted with several huge if/else branches here. Could we simplify the logic, or reverse to the previous version here?

Reply:
Thanks. Let me try improving it.